### PR TITLE
fix: harden compressor PH flashes

### PR DESCRIPTION
**Pull request by kjbrak**
## Type of Work

- [x] Patch: X.Y.**Z+1**. **NEGLIGIBLE** visible changes, does not change input or output - OR changes behaviour. Use chore:, refactor: etc
- [ ] Minor: X.**Y+1**.Z. Minor changes, might ADD new input (YAML), or other **backwards-compatible** changes. Use feat:, fix:
- [ ] Major: **X+1**.Y.Z. Major and most likely **BREAKING** changes, wo. backwards compatibility, or removing temporary backwards compatibility functionality. Use ! or BREAKING:.

See here (internal): https://github.com/equinor/ecalc-internal/discussions/1044

## Have you remembered and considered?

- [ ] IF FEAT: I have remembered to update documentation
- [x] IF FIX OR FEAT: I have remembered to update manual changelog (`docs/drafts/next.draft.md`)
- [ ] IF BREAKING: I have remembered to update migration guide (`docs/docs/migration_guides/`)
- [ ] IF BREAKING: I have committed with `BREAKING:` in footer or `!` in header
- [ ] I have added tests (if not, comment why)
- [ ] I have used conventional commits syntax (if you squash, make sure that conventional commit is used)
- [ ] I have included the Github issue nr in the footer!

## Summary
Note: BREAKING!

This PR hardens compressor PH flash handling after debugging a production compressor model crash.

The work happened in three steps:

1. **Fix the original compressor PH crash**
   - The original failure came from a compressor outlet PH flash where NeqSim returned an invalid outlet state near the phase envelope.
   - That invalid state contained bad thermodynamic properties, including non-finite `kappa`.
   - eCalc then used those properties in the Campbell pressure iteration, which produced a non-finite outlet pressure and eventually crashed in NeqSim.
   - The fix is to seed compressor PH flashes from the compressor inlet temperature instead of starting from the cached standard-condition reference state.

2. **Add validation so bad PH results cannot silently propagate**
   - PH flash results are now validated before they are cached or used.
   - Validation checks finite/positive pressure, temperature, density, `z`, `kappa`, standard density, finite vapor fraction and enthalpy, and that the resulting enthalpy matches the PH target within tolerance.
   - Compressor outlet calculations now also reject non-finite Campbell pressure guesses before they are sent back into NeqSim.

3. **Handle additional invalid/non-physical states exposed by validation**
   - After validation was added, the full model exposed another failure in a max-speed/capacity probe: the chart point was already invalid/outside capacity, but the code still required a physical outlet PH state for the extrapolated probe.
   - That PH target was not physically reachable, causing NeqSim to drive temperature toward near-zero K and fail. The code now preserves the original invalid chart/capacity result instead of crashing on the outlet flash for an already-invalid chart point.
   - Validation also exposed an existing simplified-train test case where extreme pressure ratios produced an invalid near-zero-temperature PH result. The simplified max-rate helper now avoids using those invalid PH properties once the required head is already above the maximum-speed chart head, preserving the existing chart-limited scalar result.

## Changes

- Add optional `temperature_guess_kelvin` to `FluidService.flash_ph(...)`.
- Seed compressor PH flashes from inlet stream temperature.
- Validate NeqSim PH flash results before cache insertion/use.
- Add `CompressorOutletCalculationError` for compressor outlet thermodynamic failures.
- Reject invalid/non-finite compressor PH results and Campbell pressure guesses.
- In legacy compressor evaluation, skip failed outlet PH only when the chart point is already invalid; valid chart points still raise.
- Preserve simplified max-rate behavior for above-chart-head estimates without relying on invalid NeqSim properties.

## Testing

Added and updated tests for seeded PH flash cache behavior, compressor outlet PH validation, invalid-chart outlet failure handling, and simplified max-rate behavior for invalid high-head PH states.

All tests pass.

## Verification

- All UAT models run successfully with no validation triggers — the new guards are targeted and do not affect existing production models.
- Unified Java bridge error handling: PH/TP flash handlers now catch both Py4J and JPype exceptions via a shared `JAVA_ERRORS` tuple.

## Githus projects issue
refs.:
https://github.com/equinor/ecalc-internal/issues/1765

### CODE REVIEW

File: src/ecalc_neqsim_wrapper/fluid_service.py [hunk 1]
```diff
@@ -36,6 +40,8 @@
 # Standard conditions for reference fluids and standard density calculations
 _STANDARD_TEMPERATURE_KELVIN = 288.15
 _STANDARD_PRESSURE_BARA = 1.01325
+_PH_ENTHALPY_REL_TOLERANCE = 1e-4
```
Comment (tj098895) [inline src/ecalc_neqsim_wrapper/fluid_service.py]: What are the rationales behind these tolerances? From literature? Testing? Would it be important to document this for some expert users?

File: src/ecalc_neqsim_wrapper/fluid_service.py [hunk 2]
```diff
@@ -218,6 +226,48 @@ def _make_ph_cache_key(
...
+            raise IllegalStateException(f"{context}: expected finite positive {name}, got {value}.")
+
+    @classmethod
+    def _validate_ph_result(
+        cls,
+        properties: FluidProperties,
+        target_enthalpy: float,
+        context: str,
+    ) -> None:
+        cls._require_positive_finite(properties.pressure_bara, "pressure_bara", context)
```
Comment (tj098895) [inline src/ecalc_neqsim_wrapper/fluid_service.py]: could these checks be moved to `__post__init__` for `FluidProperties` instead? Possibly add custom or Pydantic validation if that is what we want?

File: src/ecalc_neqsim_wrapper/fluid_service.py [hunk 3]
```diff
@@ -218,6 +226,48 @@ def _make_ph_cache_key(
...
+    @staticmethod
+    def _require_positive_finite(value: float, name: str, context: str) -> None:
+        if not math.isfinite(value) or value <= 0:
+            raise IllegalStateException(f"{context}: expected finite positive {name}, got {value}.")
+
+    @classmethod
+    def _validate_ph_result(
+        cls,
+        properties: FluidProperties,
+        target_enthalpy: float,
```
Comment (tj098895) [inline src/ecalc_neqsim_wrapper/fluid_service.py]:
```suggestion
        target_enthalpy_joule_per_kg: float,
```

to be consistent?

File: src/ecalc_neqsim_wrapper/fluid_service.py [hunk 4]
```diff
@@ -311,6 +361,7 @@ def flash_ph(
         fluid_model: FluidModel,
         pressure_bara: float,
         target_enthalpy: float,
+        temperature_guess_kelvin: float | None = None,
```
Comment (tj098895) [inline src/ecalc_neqsim_wrapper/fluid_service.py:318]: is it the user that should decide this? or us? not sure if I understand when it should happen, and when it should not happen?

Comment (tj098895) [reply src/ecalc_neqsim_wrapper/fluid_service.py:318]: will this have (big) consequences on results from neqsim?

File: src/ecalc_neqsim_wrapper/fluid_service.py [hunk 5]
```diff
@@ -324,26 +375,59 @@ def flash_ph(
...
+                    f"pressure_bara={pressure_bara}, target_enthalpy_joule_per_kg={target_enthalpy}, "
+                    f"temperature_guess_kelvin={temperature_guess_kelvin}, eos_model={fluid_model.eos_model}."
+                ) from error
+        else:
+            try:
+                flashed = ref.set_new_pressure_and_enthalpy(
+                    new_pressure=pressure_bara,
+                    new_enthalpy_joule_per_kg=target_enthalpy,
+                )
+            except (Py4JJavaError, Py4JError) as error:
```
Comment (tj098895) [inline src/ecalc_neqsim_wrapper/fluid_service.py]: nice to catch these errors. Are those the only errors we can get, all neqsim errors wrapped in either one of them?

however, here we are slightly violating the rule about this code being completely ignorant about py4j, neqsim, jpype etc. Could we just change to catch a generic exception, or can we request that the "reference fluid implementation" HAS TO throw an eCalc exception instead? If we change implementation etc, these exceptions will suddenly not exist any more.

File: src/libecalc/domain/process/compressor/core/train/simplified_train/simplified_train.py [hunk 6]
```diff
@@ -279,7 +280,34 @@ def calculate_maximum_rate_for_stage(
             enthalpy_change_joule_per_kg = polytropic_head / polytropic_efficiency
 
             target_enthalpy = inlet_stream.enthalpy_joule_per_kg + enthalpy_change_joule_per_kg
-            props = fluid_service.flash_ph(inlet_stream.fluid_model, outlet_pressure, target_enthalpy)
+            try:
+                props = fluid_service.flash_ph(
+                    inlet_stream.fluid_model,
```
Comment (tj098895) [inline src/libecalc/domain/process/compressor/core/train/simplified_train/simplified_train.py]: kw_args only to make sure we get it right?

File: src/libecalc/domain/process/compressor/core/train/simplified_train/simplified_train.py [hunk 7]
```diff
@@ -279,7 +280,34 @@ def calculate_maximum_rate_for_stage(
...
+                    maximum_head_on_maximum_speed_curve,
+                    exc_info=True,
+                )
+                maximum_actual_volume_rate = float(
+                    maximum_rate_function(
+                        heads=polytropic_head,  # type: ignore[arg-type]
+                        extrapolate_heads_below_minimum=False,
+                    )
+                )
+                break
```
Comment (tj098895) [inline src/libecalc/domain/process/compressor/core/train/simplified_train/simplified_train.py:321]: a bit difficult to see what will happen here imo

File: src/libecalc/domain/process/compressor/core/train/utils/common.py [hunk 8]
```diff
@@ -14,9 +18,78 @@
 RECIRCULATION_BOUNDARY_TOLERANCE = 1e-6
 # -----------------------------------------------------------------------------
 # Impossible pressure sanity cap.  Protects the PH flash when the first
-# Campbell estimate is evaluated for an artificial max‑speed head in edge cases.
+# Campbell estimate is evaluated for an artificial max-speed head in edge cases.
 # -----------------------------------------------------------------------------
 MAX_FIRST_GUESS_BAR = 2_000.0  # [bara]
+COMPRESSOR_PH_ENTHALPY_REL_TOLERANCE = 1e-4
```
Comment (tj098895) [inline src/libecalc/domain/process/compressor/core/train/utils/common.py]: relevant to reuse both the constants and the util to test for positive+finite here instead of duplicating to fluid_service? or do you not want fluid service to dep on common?

File: src/libecalc/domain/process/compressor/core/train/utils/common.py [hunk 9]
```diff
@@ -1,8 +1,12 @@
+import math
+
+from libecalc.common.errors.exceptions import IllegalStateException
```
Comment (tj098895) [inline src/libecalc/domain/process/compressor/core/train/utils/common.py]: I didnt know we had this exception. It is a bit misleadign to me, coming from Java. Then it means that the class/method has been called in an inproper way. Could we change to so something more domain related? Which state are we talking about? fluid/gas state?

File: src/libecalc/domain/process/compressor/core/train/utils/common.py [hunk 10]
```diff
@@ -14,9 +18,78 @@
...
+) -> FluidProperties:
+    try:
+        return fluid_service.flash_ph(
+            inlet_stream.fluid_model,
+            outlet_pressure_bara,
+            target_enthalpy,
+            temperature_guess_kelvin=inlet_stream.temperature_kelvin,
+        )
+    except IllegalStateException as error:
+        raise CompressorOutletCalculationError(f"{context}: PH flash failed.") from error
```
Comment (tj098895) [inline src/libecalc/domain/process/compressor/core/train/utils/common.py]: I like this way of handling exceptions, more context and better fit msg/exception for more tailored handling

File: src/ecalc_neqsim_wrapper/fluid_service.py (hunk 4 above)
Comment (kjbrak) [reply src/ecalc_neqsim_wrapper/fluid_service.py:318]: We decide to use it when we have a good guess, when we know a temperature closer to the target than standard temp.

Comment (kjbrak) [reply src/ecalc_neqsim_wrapper/fluid_service.py:318]: There is a description of it in the docstring

File: src/ecalc_neqsim_wrapper/fluid_service.py (hunk 5 above)
Comment (kjbrak) [reply src/ecalc_neqsim_wrapper/fluid_service.py]: Good comment. I will check.

File: src/libecalc/domain/process/compressor/core/train/utils/common.py (hunk 8 above)
Comment (kjbrak) [reply src/libecalc/domain/process/compressor/core/train/utils/common.py]: Good feedback, i will sort this out.

File: src/libecalc/domain/process/compressor/core/train/utils/common.py (hunk 9 above)
Comment (kjbrak) [reply src/libecalc/domain/process/compressor/core/train/utils/common.py]: I saw it had been used many places and it seemed to fit ok, but i agree, will take a look.

File: tests/ecalc_neqsim_wrapper/unit_tests/test_seeded_ph_flash.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
+from __future__ import annotations
+
+from ecalc_neqsim_wrapper.cache_service import CacheService
+from ecalc_neqsim_wrapper.fluid_service import NeqSimFluidService
+from libecalc.process.fluid_stream.fluid_model import EoSModel, FluidComposition, FluidModel
+from libecalc.process.fluid_stream.fluid_properties import FluidProperties
+
+
+class FakeNeqsimFluid:
```
Comment (tj098895) [inline tests/ecalc_neqsim_wrapper/unit_tests/test_seeded_ph_flash.py]: FakeFluid?

**Comment by tj098895** [comment]
just curious, since you didnt check the "breaking" checkbox, but indicating with **!** that it is breaking, what is breaking?

File: src/ecalc_neqsim_wrapper/fluid_service.py (hunk 1 above)
Comment (kjbrak) [reply src/ecalc_neqsim_wrapper/fluid_service.py]: Centralized the PH flash validation tolerances and documented that these are defensive backend-result checks, not process-model tolerances.

File: src/ecalc_neqsim_wrapper/fluid_service.py (hunk 2 above)
Comment (kjbrak) [reply src/ecalc_neqsim_wrapper/fluid_service.py]: Kept this at the PH flash operation boundary since validation needs the target enthalpy and operation context, which FluidProperties does not have.

File: src/ecalc_neqsim_wrapper/fluid_service.py (hunk 4 above)
Comment (kjbrak) [reply src/ecalc_neqsim_wrapper/fluid_service.py:318]: Clarified the docstring: the temperature guess is only a convergence aid for the same PH target state and is intentionally not part of the cache key.

File: src/ecalc_neqsim_wrapper/fluid_service.py (hunk 5 above)
Comment (kjbrak) [reply src/ecalc_neqsim_wrapper/fluid_service.py]: Moved backend-specific flash failures behind the NeqSim wrapper boundary and expose fluid-service flash errors to domain code instead.

File: src/libecalc/domain/process/compressor/core/train/simplified_train/simplified_train.py (hunk 7 above)
Comment (kjbrak) [reply src/libecalc/domain/process/compressor/core/train/simplified_train/simplified_train.py:321]: Refactored this path to make the chart-limited fallback explicit and only use it when the PH flash fails above chart maximum head.

File: src/libecalc/domain/process/compressor/core/train/utils/common.py (hunk 8 above)
Comment (kjbrak) [reply src/libecalc/domain/process/compressor/core/train/utils/common.py]: Moved the shared PH validation and tolerances out of compressor/common duplication so fluid service and compressor code reuse the same validation logic.

File: src/libecalc/domain/process/compressor/core/train/utils/common.py (hunk 9 above)
Comment (kjbrak) [reply src/libecalc/domain/process/compressor/core/train/utils/common.py]: Replaced this with compressor-domain thermodynamic errors carrying operation context and physical inputs.

File: src/libecalc/domain/process/compressor/core/train/utils/common.py [hunk 12]
```diff
@@ -14,9 +18,78 @@
...
+class CompressorOutletCalculationError(IllegalStateException):
+    """Raised when compressor outlet thermodynamics cannot produce a usable state."""
+
+
+def _require_positive_finite(value: float, name: str, context: str) -> None:
+    if not math.isfinite(value) or value <= 0:
+        raise CompressorOutletCalculationError(f"{context}: expected finite positive {name}, got {value}.")
+
+
+def _validate_compressor_ph_flash_result(
```
Comment (olelod) [inline src/libecalc/domain/process/compressor/core/train/utils/common.py]: Doesn't this and `_validate_ph_result` in `fluid_service.py` basically validate the same thing? The only extra here is the vapor fraction inside [0,1]. Move that to `_validate_ph_result` and drop this validation?

File: src/ecalc_neqsim_wrapper/fluid_service.py [hunk 13]
```diff
@@ -324,26 +375,59 @@ def flash_ph(
...
+            try:
+                seeded_fluid = ref.set_new_pressure_and_temperature(
+                    new_pressure_bara=pressure_bara,
+                    new_temperature_kelvin=temperature_guess_kelvin,
+                )
+                flashed = seeded_fluid.set_new_pressure_and_enthalpy(
+                    new_pressure=pressure_bara,
+                    new_enthalpy_joule_per_kg=target_enthalpy,
+                )
+            except (Py4JJavaError, Py4JError) as error:
```
Comment (olelod) [inline src/ecalc_neqsim_wrapper/fluid_service.py]: Did we drop JPype? Can errors there still leak through?

File: src/libecalc/domain/process/compressor/core/train/utils/common.py [hunk 14]
```diff
@@ -14,9 +18,78 @@
 RECIRCULATION_BOUNDARY_TOLERANCE = 1e-6
 # -----------------------------------------------------------------------------
 # Impossible pressure sanity cap.  Protects the PH flash when the first
-# Campbell estimate is evaluated for an artificial max‑speed head in edge cases.
+# Campbell estimate is evaluated for an artificial max-speed head in edge cases.
 # -----------------------------------------------------------------------------
 MAX_FIRST_GUESS_BAR = 2_000.0  # [bara]
```
Comment (olelod) [inline src/libecalc/domain/process/compressor/core/train/utils/common.py:26]: Have you checked if this one is still needed now?

File: src/libecalc/process/fluid_stream/fluid_service.py [hunk 15]
```diff
@@ -66,6 +67,12 @@ def flash_ph(
             fluid_model: The fluid model (composition + EoS)
             pressure_bara: Target pressure in bara
             target_enthalpy: Target specific enthalpy in J/kg (must be from same EoS session)
+            temperature_guess_kelvin: Optional initial temperature for the PH flash.
+                Implementations may use this to precondition the underlying thermodynamic
+                system at (pressure_bara, temperature_guess_kelvin) before solving PH.
+                Usfeul if knowing a temperature closer to final state than the reference fluid
```
Comment (olelod) [inline src/libecalc/process/fluid_stream/fluid_service.py]: Useful

File: src/ecalc_neqsim_wrapper/fluid_service.py (hunk 13 above)
Comment (tj098895) [reply src/ecalc_neqsim_wrapper/fluid_service.py]: specific implementation errors should be handled in the specific implementation (wrapper). JPype is atm inactive, but it is still there and can be configured to be used.

File: src/libecalc/domain/process/compressor/core/train/utils/common.py (hunk 12 above)
Comment (kjbrak) [reply src/libecalc/domain/process/compressor/core/train/utils/common.py]: @olelod : I did come changes yesterday before leaving the office and forgot to push, so the state this morning has been stale. New version pushed now. This issue has been resolved as the validation logic has been consolidated to `src/libecalc/process/fluid_stream/fluid_property_validation.py`

File: src/libecalc/domain/process/compressor/core/train/utils/common.py (hunk 14 above)
Comment (kjbrak) [reply src/libecalc/domain/process/compressor/core/train/utils/common.py:26]: I would think its still needed? Is there anything we have done that would make the first gas in the campbell iterations not risk being very high and crash eos liek we saw earlier? I dont think this PR changes anything regarding that.

File: src/ecalc_neqsim_wrapper/fluid_service.py (hunk 13 above)
Comment (kjbrak) [reply src/ecalc_neqsim_wrapper/fluid_service.py]: Good catch. The catch blocks in thermo.py only caught Py4JError/Py4JJavaError. JPype JException was not handled, so Java exceptions would leak through when using the JPype backend. Fixed in e8dd37f3: added a unified JAVA_ERRORS tuple in exceptions.py covering both Py4JError and JException. PH/TP flash handlers now catch both backends. Added 5 tests.

**Review by olelod — APPROVED** [review]
Should be good to go now.

Either make it a fix! or remove the Note Breaking (i think removing the note about breaking is most correct - since there is no change in input/output, and only edge cases are affected and treated more correctly)

File: .basedpyright/baseline.json [hunk 16]
```diff
@@ -393,38 +393,6 @@
                     "lineCount": 1
                 }
             },
-            {
```
Comment (olelod) [inline .basedpyright/baseline.json:396]: Should this be here?

Comment (kjbrak) [reply .basedpyright/baseline.json:396]: this file was updated automatically by pyright when i fixed the type error.. hmm

Comment (kjbrak) [reply .basedpyright/baseline.json:396]: should probably not edit it manually?

Comment (kjbrak) [reply .basedpyright/baseline.json:396]: Looking at it, what is the problem here really, i dont see it.

**Review by jsolaas — APPROVED** [review]
